handler/src/main/java/io/netty/handler/ssl/OptionalSslHandler.java

```diff
@@ -0,0 +1,111 @@
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.internal.ObjectUtil;
+
+import javax.net.ssl.SSLContext;
+import java.util.List;
+
+/**
+ * {@link OptionalSslHandler} is a utility decoder to support both SSL and non-SSL handlers
+ * based on the initial message.
+ */
+public class OptionalSslHandler extends ByteToMessageDecoder {
+
+    private final SslContext sslContext;
+
+    public OptionalSslHandler(SslContext sslContext) {
+        this.sslContext = ObjectUtil.checkNotNull(sslContext, "sslContext");
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext context, ByteBuf in, List<Object> out) throws Exception {
+        if (in.readableBytes() < SslUtils.SSL_RECORD_HEADER_LENGTH) {
+            return;
+        }
+        if (SslHandler.isEncrypted(in)) {
+            handleSsl(context);
+        } else {
+            handleNonSsl(context);
+        }
+    }
+
+    private void handleSsl(ChannelHandlerContext context) {
+        SslHandler sslHandler = null;
+        try {
+            sslHandler = newSslHandler(context, sslContext);
+            context.pipeline().replace(this, newSslHandlerName(), sslHandler);
+            sslHandler = null;
+        } finally {
+            // Since the SslHandler was not inserted into the pipeline the ownership of the SSLEngine was not
+            // transferred to the SslHandler.
+            if (sslHandler != null) {
+                ReferenceCountUtil.safeRelease(sslHandler.engine());
+            }
+        }
+    }
+
+    private void handleNonSsl(ChannelHandlerContext context) {
+        ChannelHandler handler = newNonSslHandler(context);
+        if (handler != null) {
+            context.pipeline().replace(this, newNonSslHandlerName(), handler);
+        } else {
+            context.pipeline().remove(this);
+        }
+    }
+
+    /**
+     * Optionally specify the SSL handler name, this method may return {@code null}.
+     * @return the name of the SSL handler.
+     */
+    protected String newSslHandlerName() {
+        return null;
+    }
+
+    /**
+     * Override to configure the SslHandler eg. enable HttpsEndpointIdentificationAlgorithm.
```

Comment:
Add a comment to the effect: `the hostname and port is not known by this method so servers may want to override this method and use the {@link SslHandler#newHandlerByteBufAllocator, string, int)} variant.`

Also replace `HttpsEndpointIdentificationAlgorithm` with `{@link SSLParamters#setEndpointIdentificationAlgorithm(String)}`